### ISSUE TITLE
PatchDetour::process_ saves pointer to by-reference variable

### ISSUE THREAD

**Issue by namreeb** [RaptorFactor/hadesmem#7]
This is bad because there is no obvious hint to the user that you should not pass a local stack variable to the PatchDetour constructor.  It would be better if process_ were a copy of the parameter to the constructor, rather than a pointer to it.

**Comment by Maktm** [comment]
I'm assuming @RaptorFactor designed it this way because constructing/copying hadesmem::Process objects are expensive. Of course it comes at the cost of the user accidentally passing in an object that goes out of scope, but it could always be specified in the docs (which is part of the TODO). Also note that PatchDetour isn't the only component that does this but also all the other components in hadesmem that require a hadesmem::Process object. Users can always do what @RaptorFactor did in the examples, which is:

    hadesmem::Process& GetThisProcess()
    {
      static hadesmem::Process process(::GetCurrentProcessId());
      return process;
    }

**Comment by namreeb** [comment]
I don't think the cost of constructing a hadesmem::Process object is the issue.  I think the issue is that there is nothing about the code which clearly indicates that the lifetime of the hadesmem::Process object must be greater than the lifetime of the hadesmem::PatchDetour object.  In fact, the opposite is the heavily favored intuitive design.  Using a pointer rather than a reference (and optionally a by-value copy) still allows the user to avoid future instantiations of the hadesmem::Process type.

**Comment by RaptorFactor** [comment]
I will be back from the dead some-time this week. I'm still being randomized by dumb shit (immigration is fun).

**Comment by RaptorFactor** [comment]
In the general case I agree it's unfortunate that there's nothing obvious to indicate that most types rely on having the process object outlive them. In this case it might make sense for the Patcher to own the process because unlike most typical use cases for other types, in the case of detours you're often initializing the Patcher object in a context that has nothing to do with its lifetime.

I'll think about it and do some testing, because I'm pretty sure that there are some issues with copying Process objects when injected into games protected by certain anti-cheats with a kernel-mode component filtering process access. I think I already solved that by using pseudo-handles in that case, but I still need to verify.

**Comment by namreeb** [comment]
Could always do a shared_ptr

**Comment by RaptorFactor** [comment]
Right now the Patcher constructors which take an rvalue Process object are deleted (specifically to try and prevent accidental misuse similar to but not exactly like this case), so what if I changed the code so that if you pass in an rvalue then the Patcher object will take ownership of it, otherwise it will keep the existing behavior? That way it should be a completely non-breaking change, and users who want the new behavior can do:
Process proc(blah);
Patcher patch(std::move(proc), ...);

Thoughts?

shared_ptr is indeed another option that shouldn't break current uses (I did some basic tests against hardened processes). It does break the interface symmetry, although the default interface is indeed not ideal so for the sake of pragmatism I'm not opposed to breaking it in this case if you prefer that solution.
